webkitgtk: refactor top-level to deprecate default ABI version

fabianhjr opened the pull request:
## Description of changes

This is to avoid breaking things by changing the default version and making dependents more explicit about such dependency

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fabianhjr committed:
webkitgtk: refactor top-level to deprecate default ABI version

fabianhjr commented on pkgs/top-level/all-packages.nix:
```diff
@@ -23698,21 +23698,23 @@ with pkgs;
     gtk3 = gtk4;
   };
 
+  webkitgtk = webkitgtk_4_0;
```
Unsure if this would be preferable to move to aliases.nix with a warn to explicitly set the ABI version instead of relying on whichever is provided by default. (I think this would be better)

khaneliman replied:
I feel like I've seen differing opinions on this as to whether to encourage version pinning or not, to avoid having to maintain many versions of programs.

fabianhjr replied:
Another thing is that the three abi versions (4_0, 4_1, 6_0) are on the same version of webkitgtk, those suffixes don't indicate the version of webkitgtk being used.

fabianhjr replied:
Merging this and opening a different PR suggesting a move to a warn.

fabianhjr replied:
Amending this

fabianhjr committed:
treewide: replace webkitgtk to webkitgtk_4_0

sed -i 's/ webkitgtk\b/ webkitgtk_4_0/g' pkgs/**.nix
sed -i 's/(webkitgtk\b/(webkitgtk_4_0/g' pkgs/**.nix
sed -i 's/\.webkitgtk\b/.webkitgtk_4_0/g' pkgs/**.nix

webkitgtk is currently pointing to that specific ABI version but the
alias is going to start warning

fabianhjr commented:
```
1 package added:
webkitgtk_4_0 (init at 2.46.0+abi=4.0)

1 package removed:
webkitgtk (†2.46.0+abi=4.0)

4 packages built:
webkitgtk_4_0 webkitgtk_4_0.debug webkitgtk_4_0.dev webkitgtk_4_0.devdoc
```

fabianhjr commented:
Result of `nixpkgs-review pr 345611` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.debug</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

fabianhjr commented:
Cleaned up the tree @jtojnar, will wait a bit for feedback but this seems mergeable as is.

khaneliman commented:
```bash
 error: function 'anonymous lambda' called without required argument 'webkitgtk_4_0-web-process-extension'
       at /ofborg/checkout/2/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-1/pkgs/development/haskell-modules/hackage-packages.nix:124272:6:
        124271|   "gi-webkitwebprocessextension" = callPackage
        124272|     ({ mkDerivation, base, bytestring, Cabal, containers, gi-gdk
             |      ^
        124273|      , gi-gio, gi-gobject, gi-gtk, gi-javascriptcore, gi-soup

khaneliman commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 345611`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.debug</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.debug</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

sternenseemann commented:
`hackage-packages.nix` is generated automatically. Code generation is done by [cabal2nix](https://github.com/nixos/cabal2nix) which would need to be adjusted if anything. I don't think changing generation there is a good thing. cabal2nix generally just resolves names and requires the user to pass in a compatible version of each dependency, so it would be bizarre for it to decide that every webkit user shall henceforth pull in webkitgtk 4.

fabianhjr commented:
Created https://github.com/NixOS/cabal2nix/pull/629

Addressed the eval failure

fabianhjr commented:
Result of `nixpkgs-review pr 345611` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.debug</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

fabianhjr commented:
There was a merge conflict, rebased

jtojnar commented on pkgs/development/libraries/webkit2-sharp/default.nix:
```diff
@@ -45,7 +45,7 @@ stdenv.mkDerivation rec {
 
   passthru = {
-    inherit webkitgtk;
+    inherit webkitgtk_4_0;
```
This should probably remain so that the user can get the correct ABI version used by this library. Otherwise, they could just use the top-level one.

jtojnar commented on pkgs/top-level/aliases.nix:
```diff
@@ -1670,6 +1670,7 @@ mapAliases {
   waypoint = throw "waypoint has been removed from nixpkgs as the upstream project was archived"; # Added 2024-04-24
   wcm = throw "'wcm' has been renamed to/replaced by 'wayfirePlugins.wcm'"; # Add 2023-07-29
   webkitgtk_5_0 = throw "'webkitgtk_5_0' has been superseded by 'webkitgtk_6_0'"; # Added 2023-02-25
+  webkitgtk = lib.warn "Explicitly set the ABI version of 'webkitgtk'" webkitgtk_4_0;
```
Moving it to aliases without getting rid of them from  tree will break eval with `!allowAliases` on the commit. This might be annoying e.g. when bisecting. Either keep the `webkitgtk = webkitgtk_4_0` in `all-packages.nix` in the first commit and change it to aliases in the second one, or squash the commits.

fabianhjr commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29251,7 +29251,6 @@ with pkgs;
 
   font-manager = callPackage ../by-name/fo/font-manager/package.nix {
     libsoup = libsoup_3;
```
Changed in by-name and removed the entry in top-level/all-packages.nix

fabianhjr commented:
Addressed all your comments @jtojnar, thanks for the review. :3

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30847,7 +30842,6 @@ with pkgs;
 
   liferea = callPackage ../applications/networking/newsreaders/liferea {
     libsoup = libsoup_3;
```
Same issue as with the font-manager.

jtojnar commented on pkgs/development/libraries/webkit2-sharp/default.nix:
Looks like this broke the eval.

fabianhjr replied:
looking into this

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -23582,17 +23582,17 @@ with pkgs;
 
   wcslib = callPackage ../development/libraries/science/astronomy/wcslib { };
 
-  webkitgtk = callPackage ../development/libraries/webkitgtk {
```
Now the first commit does not evaluate. It references `webkitgtk_4_0` but it is only introduced in the second one.

fabianhjr replied:
First commit is now introducing both webkitgtk_4_0 and aliasing on all-packages.nix webkitgtk to webkitgtk_4_0.

Then treewide changes

Finally moving webkitgtk alias into aliases with a warning.

fabianhjr committed:
webkitgtk: move webkitgtk to aliases with a warning

fabianhjr commented:
Working on rebase

jtojnar reviewed (approved):
Thanks, LGTM now.

fabianhjr commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 345611`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>webkitgtk_4_0</li>
    <li>webkitgtk_4_0.debug</li>
    <li>webkitgtk_4_0.dev</li>
    <li>webkitgtk_4_0.devdoc</li>
  </ul>
</details>

RossComputerGuy reviewed (approved):
Quick skim through, LGTM.

K900 reviewed (approved):
LGTM conceptually, want to see 0 rebuilds confirmed.

fabianhjr commented:
Detects the webkitgtk_4_0 as a new package so hasn't reported 0. :c

Labelled as 1 rebuild on previous evals

fabianhjr commented:
@ofborg eval

fabianhjr commented:
Eval finished as 1 rebuild due to webkitgtk_4_0 being treated as a new package
